### PR TITLE
Silence some wrong unsafe-member-access errors

### LINKED ISSUE #87: TypeScript type-aware checks don't work in template expressions or cross-context
It appears that with TypeScript support on (and type-aware checks enabled), values are (*sometimes* - see below) erroneously inferred to have `any` type in 2 situations:
1. Values declared in the script block and used in template expressions
2. Values declared in context script blocks and used in instance script blocks

Here's an script with a various examples:

<details>
  <summary><b><em>exports.ts</em></b></summary>

  ```ts
  export const import_foo1 = {
      bar: 4,
      length: 5
  };

  export const import_foo2 = [ import_foo1.bar ];
  ```
</details>

**_Demo.svelte_**:
```svelte
<script lang="ts" context="module">
	const context_foo1 = { bar: 5 };
	const context_foo2 = [ context_foo1.bar ]; // OK
	console.log(context_foo2.length); // OK
</script>

<script lang="ts">
	import { createEventDispatcher } from "svelte";
	import {import_foo1, import_foo2} from "./exports";

	export let foo1: string[];
	export let foo2: { length: number; };
	export let foo3: { bar: number; };

	const foo4 = { length: 4 };
	const foo5 = {
		length: 4,
		bar: 5,
	};
	const foo6 = function(){};
	const dispatch = createEventDispatcher();

	// All OK
	console.log(foo1.length);
	console.log(foo2.length);
	console.log(foo3.bar);
	console.log(foo4.length);
	console.log(foo5.length);
	console.log(foo5.bar);
	dispatch("event");
	console.log(import_foo1.length);
	console.log(import_foo1.bar);
	console.log(import_foo2.length);

	// OK with strict
	console.log(context_foo1.bar);

	// Unsafe member access .length on an any value. (@typescript-eslint/no-unsafe-member-access)
	console.log(context_foo2.length);
</script>

{foo1.length} <!-- @typescript-eslint/no-unsafe-member-access -->
{foo2.length} <!-- @typescript-eslint/no-unsafe-member-access -->
{foo3.bar} <!-- OK with strict -->
{foo4.length} <!-- @typescript-eslint/no-unsafe-member-access -->
{foo5.length} <!-- @typescript-eslint/no-unsafe-member-access -->
{foo5.bar} <!-- OK with strict -->
{import_foo1.length} <!-- @typescript-eslint/no-unsafe-member-access -->
{import_foo1.bar} <!-- OK with strict -->
{import_foo2.length} <!-- @typescript-eslint/no-unsafe-member-access -->

<!--Unsafe call of an any typed value (@typescript-eslint/no-unsafe-call) -->
<button on:click={() => dispatch("event")}>Click Me</button>

<!--Unsafe call of an any typed value (@typescript-eslint/no-unsafe-call) -->
<button on:click={() => foo6()}>Click Me Baby One More Time!</button>

{context_foo1.bar} <!-- OK with strict -->
{context_foo2.length} <!-- @typescript-eslint/no-unsafe-member-access -->
```

In my `tsconfig.json`, if I set `strict` to `true`, then the lines marked "OK with strict" above. If I set `strict` to `false`, they error with `no-unsafe-member-access`. All other commented lines appear to fail regardless.

So, there seems to be something weird about the way TypeScript treats the `length` property (and maybe others - I haven't investigated), but either way, **I believe all these eslint errors are incorrect**.

For completeness, here's my config:

<details>
  <summary><b><em>package.json</em></b></summary>

  ```json
  {
    "name": "demo",
    "version": "1.0.0",
    "devDependencies": {
      "@tsconfig/svelte": "^1.0.10",
      "@typescript-eslint/eslint-plugin": "^4.15.1",
      "@typescript-eslint/parser": "^4.15.1",
      "eslint": "^7.20.0",
      "eslint-plugin-svelte3": "^3.1.1",
      "svelte": "^3.32.3",
      "svelte-preprocess": "^4.6.9",
      "typescript": "^4.1.5"
    },
    "scripts": {
      "lint": "eslint --ext .svelte,.ts ."
    }
  }
  ```
</details>

<details>
  <summary><b><em>tsconfig.json</em></b></summary>

  ```json
  {
    "extends": "@tsconfig/svelte/tsconfig.json",
    "include": ["**/*.ts", "**/*.svelte"],
    "compilerOptions": {
      "lib": ["DOM"],
      "strict": true,
      "baseUrl": ".",
    }
  }
  ```
</details>

<details>
  <summary><b><em>.eslintrc.js</em></b></summary>

  ```js
module.exports = {
    parser: "@typescript-eslint/parser",

    env: {
      browser: true,
    },

    plugins: [
      "svelte3",
      "@typescript-eslint",
    ],

    parserOptions: {
      project: ["./tsconfig.json"],
      tsconfigRootDir: __dirname,
      extraFileExtensions: [".svelte"],
    },

    settings: {
      "svelte3/typescript": require("typescript"),
    },

    rules: {
      "@typescript-eslint/no-unsafe-member-access": "error",
      "@typescript-eslint/no-unsafe-call": "error"
    },

    overrides: [{
      files: ["**/*.svelte"],
      processor: "svelte3/svelte3",
    }],
  };
  ```
</details>

### PR DESCRIPTION
**Pull request by dummdidumm**
Fixes #87
.. but not all possible cases, see README.

### CODE REVIEW

File: test/samples/typescript/expected.json [hunk 1]
```diff
@@ -35,8 +35,8 @@
 				"messageId": "suggestUnknown",
 				"fix": {
 					"range": [
-						29,
-						32
+						58,
```
Comment (dummdidumm) [inline test/samples/typescript/expected.json:38]: These are the ranges of the suggestions. They are not yet mapped back correctly, so I had to change them here. Suggestions did not work before and do not now, so no change in functionality there.

File: src/preprocess.js [hunk 2]
```diff
@@ -209,14 +226,10 @@ const ts_import_transformer = (context) => {
 // 5. parse the source to get the AST
 // 6. return AST of step 5, warnings and vars of step 2
 function compile_code(text, compiler, processor_options) {
```
Comment (dummdidumm) [inline src/preprocess.js:228]: I refactored the code for readability, functionality stayed the same.

File: test/samples/typescript/expected.json (hunk 1 above)
Comment (hWorblehat) [reply test/samples/typescript/expected.json:38]: Does this imply that `--fix` will make bad edits to `.svelte` files? If so, shouldn't there be something in the README warning about this? I appreciate (as you said) that this isn't something new with this PR.

**Review by hWorblehat — COMMENTED** [review]
As the reporter of #87, I thought I'd take a look. To my untrained eye, the fix looks like it should work, and the additional test sample should cover it.

Just a couple of comments...

File: test/samples/typescript/expected.json (hunk 1 above)
Comment (dummdidumm) [reply test/samples/typescript/expected.json:38]: No, this is only related to suggestions. Suggestions are different than fixes. You can select one of multiple suggestions in the IDE. That feature is not supported yet, independently of TS support.

Comment (hWorblehat) [reply test/samples/typescript/expected.json:38]: OK - thanks for clarifying :+1:

**Comment by hWorblehat** [comment]
I can see this change works by effectively splicing the script blocks into the code being analysed, thereby ensuring all declarations are present as they were originally written.

I haven't looked in too much detail at how this plugin works, but is there a chance that with this change, if there is an ESLint warning in (say) the module context block, then it will be double-reported due to the module context being spliced into other blocks?

**Comment by dummdidumm** [comment]
It will not (and the test confirms that) because eslint warnings/errors on unmapped code will be ignored.

**Comment by hWorblehat** [comment]
Well I'm happy then, but presumably my approval isn't very useful in terms of getting this merged...

**Pull request merged**